### DISCUSSION TITLE
Topocentric elongation of moon.

### DISCUSSION THREAD

**Discussion by hidp123** [cosinekitty/astronomy#241] [Q&A] [answered]
Hi, using the elongation function [here](https://github.com/cosinekitty/astronomy/tree/master/source/python#Elongation) I can return the geocentric elongation of the moon (from the centre of the Earth). Is there a way to return the topocentric elongation as well (for a given location)?

Regards

**Comment by cosinekitty** [comment, marked as answer]
Hi @hidp123! Yes, I can help you with this. Here is the simplest way I can think of right now.

Call the [Equator](https://github.com/cosinekitty/astronomy/tree/master/source/python#equatorbody-time-observer-ofdate-aberration) function twice, once with `Body.Sun`, again with `Body.Moon`. In both calls, pass the desired geographic coordinates in the `observer` parameter. These calls will return type [Equatorial](https://github.com/cosinekitty/astronomy/tree/master/source/python#Equatorial), which contains a vector called `vec`. 

Then pass both vectors to [AngleBetween](https://github.com/cosinekitty/astronomy/tree/master/source/python#anglebetweena-b), which returns the angle between the vectors expressed in degrees.

Does this help?

**Reply by hidp123** [reply]
Thanks. This worked. Comparison below:

**Geocentric elongation:**
```
sunset   = astronomy.Time.Make(2022, 6, 29, 20, 29, 42)
moon_elongation = astronomy.Elongation(astronomy.Body.Moon, sunset) #geocentric elongation

print(moon_elongation.elongation)
```
This gives me 9.005544649892581
JPL Horizons gives me: 9.0052
___
**Topocentric elongation:**
```
observer = astronomy.Observer(52, -1.5)
sunset = astronomy.Time.Make(2022, 6, 29, 20, 29, 42)
sun_equator = astronomy.Equator(astronomy.Body.Sun, sunset, observer, 1, None)
moon_equator = astronomy.Equator(astronomy.Body.Moon, sunset, observer, 1, None)

elongation_topo = astronomy.AngleBetween(sun_equator.vec, moon_equator.vec)

print(elongation_topo)
```
This gives me: 8.292573166503782
JPL Horizons gives me: 8.2974

**Reply by hidp123** [reply]
A suggestion:
Add both options to the elongation function, geocentric and topocentric, depending on which argument the user provides.
Or, a separate function for elongation_topo, as more arguments are needed for it.
Or add it to the horizon function.

**Reply by cosinekitty** [reply]
Topocentric correction is definitely important for the Moon. It seems like this issue could appear for other situations also, not just elongation. So I want to think about this in more depth, to see if there is a more general way of correcting the Moon's position for different observers, without having to use the `Equator` function.

I will mention something that might be helpful in general. You can use [ObserverVector](https://github.com/cosinekitty/astronomy/tree/master/source/python#ObserverVector) to calculate a vector from the center of the Earth to a geographic location on the Earth. This vector can be subtracted from a geocentric vector to obtain a topocentric vector. So another way to do the calculation is

```python
observer = astronomy.Observer(52, -1.5)
sunset = astronomy.Time.Make(2022, 6, 29, 20, 29, 42)
sun = astronomy.GeoVector(astronomy.Body.Sun, sunset, True)
moon = astronomy.GeoMoon(sunset)
ovec = astronomy.ObserverVector(sunset, observer, False)
print(astronomy.AngleBetween(sun-ovec, moon-ovec))
```

The output is: 8.297686120182458

This is closer to the JPL Horizons calculation, and it may be slightly faster too. I think the difference is, you didn't correct for aberration of the Sun, which can make a 22 arcsecond difference. Aberration does not matter for the Moon.

**Reply by hidp123** [reply]
Thanks.
In my above code changing aberration to `true` gives the same result as well. Not sure about performance though.

```
observer = astronomy.Observer(52, -1.5)
sunset = astronomy.Time.Make(2022, 6, 29, 20, 29, 42)
sun_equator = astronomy.Equator(astronomy.Body.Sun, sunset, observer, True, True)
moon_equator = astronomy.Equator(astronomy.Body.Moon, sunset, observer, True, True)

elongation_topo = astronomy.AngleBetween(sun_equator.vec, moon_equator.vec)

print(elongation_topo)
```

Output: 8.297686120182458

**Reply by cosinekitty** [reply]
Yes, that makes sense that the answer is exactly the same either way. About performance, they are probably about the same, because the most expensive part is calculating the Moon's position. Just that one thing is probably 90% of the work for the CPU. But it's not obvious to me which way is better. So you could try putting each calculation inside a loop and repeat each 1000 times and time how long they take respectively. I have learned the hard way that measuring performance is always better than guessing!

**Comment by hidp123** [comment]
I'm a bit confused regarding the following terms:

- astrometric. My understanding: true position of the body in space.
- apparent: My understanding: position of the body as seen by an observer. So, for example, the true position of the sun 8 minutes ago.
- aberration corrected time: this is the one I'm not sure about. Is it one of the above two (mainly the second), or a third thing altogether?

(the above terms are from Skyfield python package)

**Reply by cosinekitty** [reply]
Yes, astrometric means the actual position of the body in space, without correcting for light travel time *or* aberration. Apparent positions correct for light travel time and (optionally) aberration.

Aberration is an apparent shift in the direction of an object due to perpendicular movement of the observer with respect to the incoming rays of light. The best analogy I can think of is, if it is raining and the rain is coming straight down from the sky, and I ask you to point to where the rain is coming from, you will point straight up. But if you are driving in a car, it will look like the rain is coming down at an angle. If I ask you to point to where the rain is coming from while driving, you will point diagonally upward and somewhat in the direction you are driving.

Like rain, visible light has a finite speed, and you get the same effect. The direction the light is coming from (where you see the object in the sky) can change due to movement of the Earth. As the Earth moves around the Sun, it can cause anywhere from 0 to 22 arcseconds of angular shift in the apparent location of objects. Aberration angles are largest when the observed object is perpendicular to the Earth's velocity vector, and smallest when they are parallel.

Light travel time correction is a separate issue. This is a correction for the fact that we see a planet like Jupiter where it was when the light left Jupiter on its way to the Earth, not where Jupiter is right now.

The functions `Equator` and `GeoVector` always correct for light travel time. However, they only correct for aberration when you tell them to.

The reason I give people the choice is that sometimes we may want to determine where a planet is with respect to background stars. If you see Jupiter right next to a star, both will have very similar aberration, and you might not want to correct for aberration so that it's easier to know Jupiter is close to the known star position. In another case, you might want to know relative to telescope instruments exactly what position in the sky the object appears; here you would want to correct for aberration.

Does this help?

**Reply by hidp123** [reply]
Yes, thanks.

Following on from this, is it possible to calculate the astrometric position as well on astronomy-engine ? I don't need to atm but was curious as I was comparing with Skyfield.

**Reply by cosinekitty** [reply]
Yes, you can do this by using [HelioVector](https://github.com/cosinekitty/astronomy/tree/master/source/python#HelioVector) instead, which does not correct for aberration or light travel time. You would calculate the heliocentric vector for the Earth, and the heliocentric vector for the other body. Subtract the two vectors to obtain the astrometric geocentric vector.

You can then use [EquatorFromVector](https://github.com/cosinekitty/astronomy/tree/master/source/python#EquatorFromVector) to convert the vector into equatorial angular coordinates.

**Reply by cosinekitty** [reply]
I should mention for completeness that light travel time and aberration are insignificant for the Moon, because it is so close and its transverse speed is so small. So the Moon's astrometric and apparent positions are essentially the same.
